File: server/utils/ErrorHandler.js
/**
 * Gestionnaire d'erreurs centralisé pour le serveur
 *
 * Fournit une gestion cohérente des erreurs avec logging, classification et réponses standardisées
 */

export class ErrorHandler {
  constructor(logger) {
    this.logger = logger
    this.errorCounts = new Map() // type -> count
    this.errorHistory = [] // Dernières erreurs
    this.maxHistorySize = 100
  }

  /**
   * Gère une erreur et retourne une réponse standardisée
   *
   * @param {Error} error - L'erreur à gérer
   * @param {object} context - Contexte supplémentaire
   * @param {string} context.action - Action qui a échoué
   * @param {string} context.clientId - ID du client
   * @param {object} context.payload - Payload de la requête
   * @returns {object} Réponse d'erreur standardisée
   */
  handleError(error, context = {}) {
    const errorType = this.classifyError(error)
    const errorId = this.generateErrorId()

    // Incrémenter le compteur
    this.errorCounts.set(errorType, (this.errorCounts.get(errorType) || 0) + 1)

    // Ajouter à l'historique
    this.addToHistory({
      id: errorId,
      type: errorType,
      message: error.message,
      stack: error.stack,
      context,
      timestamp: new Date().toISOString(),
    })

    // Logger selon le type
    this.logError(error, errorType, context)

    // Retourner une réponse standardisée
    return this.createErrorResponse(error, errorType, errorId, context)
  }

  /**
   * Classe une erreur selon son type
   */
  classifyError(error) {
    if (error.name === 'ValidationError') return 'validation'
    if (error.name === 'AuthenticationError') return 'authentication'
    if (error.name === 'AuthorizationError') return 'authorization'
    if (error.name === 'RateLimitError') return 'rate_limit'
    if (error.name === 'TimeoutError') return 'timeout'
    if (error.name === 'NetworkError') return 'network'
    if (error.name === 'PluginError') return 'plugin'
    if (error.message?.includes('not found')) return 'not_found'
    if (error.message?.includes('permission')) return 'permission'
    if (error.message?.includes('timeout')) return 'timeout'
    return 'unknown'
  }

  /**
   * Génère un ID unique pour l'erreur
   */
  generateErrorId() {
    return `err_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
  }

  /**
   * Ajoute une erreur à l'historique
   */
  addToHistory(errorInfo) {
    this.errorHistory.push(errorInfo)
    if (this.errorHistory.length > this.maxHistorySize) {
      this.errorHistory.shift()
    }
  }

  /**
   * Log une erreur selon son type
   */
  logError(error, errorType, context) {
    const logData = {
      type: errorType,
      message: error.message,
      action: context.action,
      clientId: context.clientId,
    }

    switch (errorType) {
      case 'validation':
        this.logger.warn('Validation error', logData)
        break
      case 'authentication':
      case 'authorization':
        this.logger.warn('Auth error', logData)
        break
      case 'rate_limit':
        this.logger.warn('Rate limit exceeded', logData)
        break
      case 'timeout':
        this.logger.warn('Timeout error', logData)
        break
      case 'network':
        this.logger.error('Network error', logData)
        break
      case 'plugin':
        this.logger.error('Plugin error', logData)
        break
      default:
        this.logger.error('Unknown error', { ...logData, stack: error.stack })
    }
  }

  /**
   * Crée une réponse d'erreur standardisée
   */
  createErrorResponse(error, errorType, errorId, context) {
    const baseResponse = {
      success: false,
      error: {
        id: errorId,
        type: errorType,
        message: this.getUserFriendlyMessage(error, errorType),
        timestamp: new Date().toISOString(),
      },
    }

    // Ajouter des détails selon le type
    switch (errorType) {
      case 'validation':
        baseResponse.error.details = error.details || {}
        baseResponse.error.field = error.field
        break
      case 'rate_limit':
        baseResponse.error.retryAfter = error.retryAfter
        break
      case 'timeout':
        baseResponse.error.timeout = error.timeout
        break
    }

    // En mode développement, ajouter la stack
    if (process.env.NODE_ENV === 'development') {
      baseResponse.error.stack = error.stack
      baseResponse.error.originalMessage = error.message
    }

    return baseResponse
  }

  /**
   * Retourne un message utilisateur-friendly
   */
  getUserFriendlyMessage(error, errorType) {
    const messages = {
      validation: 'Les données fournies sont invalides',
      authentication: 'Authentification requise',
      authorization: 'Accès non autorisé',
      rate_limit: 'Trop de requêtes, veuillez réessayer plus tard',
      timeout: 'La requête a expiré',
      network: 'Erreur de connexion réseau',
      plugin: 'Erreur dans le plugin',
      not_found: 'Ressource non trouvée',
      permission: 'Permissions insuffisantes',
      unknown: 'Une erreur inattendue s\'est produite',
    }

    return messages[errorType] || messages.unknown
  }

  /**
   * Obtient les statistiques d'erreurs
   */
  getErrorStats() {
    return {
      counts: Object.fromEntries(this.errorCounts),
      recent: this.errorHistory.slice(-10),
      total: this.errorHistory.length,
    }
  }

  /**
   * Réinitialise les statistiques
   */
  resetStats() {
    this.errorCounts.clear()
    this.errorHistory = []
  }

  /**
   * Crée une erreur typée
   */
  static createError(type, message, details = {}) {
    const error = new Error(message)
    error.name = `${type}Error`
    error.type = type
    Object.assign(error, details)
    return error
  }

  /**
   * Wrapper pour async functions avec gestion d'erreur
   */
  async wrapAsync(fn, context = {}) {
    try {
      return await fn()
    } catch (error) {
      return this.handleError(error, context)
    }
  }
}

/**
 * Erreurs personnalisées
 */
export class ValidationError extends Error {
  constructor(message, field, details = {}) {
    super(message)
    this.name = 'ValidationError'
    this.field = field
    this.details = details
  }
}

export class AuthenticationError extends Error {
  constructor(message = 'Authentication required') {
    super(message)
    this.name = 'AuthenticationError'
  }
}

export class AuthorizationError extends Error {
  constructor(message = 'Access denied') {
    super(message)
    this.name = 'AuthorizationError'
  }
}

export class RateLimitError extends Error {
  constructor(message = 'Rate limit exceeded', retryAfter = 60) {
    super(message)
    this.name = 'RateLimitError'
    this.retryAfter = retryAfter
  }
}

export class TimeoutError extends Error {
  constructor(message = 'Operation timeout', timeout = 30000) {
    super(message)
    this.name = 'TimeoutError'
    this.timeout = timeout
  }
}

export class PluginError extends Error {
  constructor(pluginName, message, details = {}) {
    super(`Plugin ${pluginName}: ${message}`)
    this.name = 'PluginError'
    this.pluginName = pluginName
    this.details = details
  }
}

File: android/app/src/main/java/com/androidcontroldeck/ui/utils/ErrorHandler.kt
package com.androidcontroldeck.ui.utils

import com.androidcontroldeck.logging.AppLogger
import androidx.compose.material3.SnackbarHostState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/**
 * Gestionnaire d'erreurs centralisé pour l'application Android
 */
class ErrorHandler(
    private val scope: CoroutineScope,
    private val snackbarHostState: SnackbarHostState? = null
) {
    private val errorHistory = mutableListOf<ErrorInfo>()
    private val maxHistorySize = 50

    /**
     * Gère une erreur et affiche un message à l'utilisateur
     */
    fun handleError(
        error: Throwable,
        context: ErrorContext = ErrorContext(),
        showSnackbar: Boolean = true
    ) {
        val errorInfo = ErrorInfo(
            type = classifyError(error),
            message = error.message ?: "Erreur inconnue",
            stackTrace = error.stackTraceToString(),
            context = context,
            timestamp = System.currentTimeMillis()
        )

        // Ajouter à l'historique
        addToHistory(errorInfo)

        // Logger
        logError(error, errorInfo)

        // Afficher un snackbar si demandé
        if (showSnackbar && snackbarHostState != null) {
            scope.launch {
                snackbarHostState.showSnackbar(
                    message = getUserFriendlyMessage(errorInfo),
                    duration = androidx.compose.material3.SnackbarDuration.Short
                )
            }
        }
    }

    /**
     * Classe une erreur selon son type
     */
    private fun classifyError(error: Throwable): ErrorType {
        return when {
            error is java.net.UnknownHostException -> ErrorType.NETWORK
            error is java.net.SocketTimeoutException -> ErrorType.TIMEOUT
            error is java.io.IOException -> ErrorType.IO
            error.message?.contains("authentication", ignoreCase = true) == true -> ErrorType.AUTHENTICATION
            error.message?.contains("permission", ignoreCase = true) == true -> ErrorType.PERMISSION
            error.message?.contains("not found", ignoreCase = true) == true -> ErrorType.NOT_FOUND
            error.message?.contains("validation", ignoreCase = true) == true -> ErrorType.VALIDATION
            else -> ErrorType.UNKNOWN
        }
    }

    /**
     * Ajoute une erreur à l'historique
     */
    private fun addToHistory(errorInfo: ErrorInfo) {
        errorHistory.add(errorInfo)
        if (errorHistory.size > maxHistorySize) {
            errorHistory.removeAt(0)
        }
    }

    /**
     * Log une erreur
     */
    private fun logError(error: Throwable, errorInfo: ErrorInfo) {
        when (errorInfo.type) {
            ErrorType.NETWORK -> AppLogger.w(TAG, "Network error: ${errorInfo.message}", error)
            ErrorType.TIMEOUT -> AppLogger.w(TAG, "Timeout error: ${errorInfo.message}", error)
            ErrorType.AUTHENTICATION -> AppLogger.w(TAG, "Auth error: ${errorInfo.message}", error)
            ErrorType.PERMISSION -> AppLogger.w(TAG, "Permission error: ${errorInfo.message}", error)
            else -> AppLogger.e(TAG, "Error: ${errorInfo.message}", error)
        }
    }

    /**
     * Retourne un message utilisateur-friendly
     */
    private fun getUserFriendlyMessage(errorInfo: ErrorInfo): String {
        return when (errorInfo.type) {
            ErrorType.NETWORK -> "Erreur de connexion réseau"
            ErrorType.TIMEOUT -> "La requête a expiré"
            ErrorType.IO -> "Erreur de lecture/écriture"
            ErrorType.AUTHENTICATION -> "Authentification requise"
            ErrorType.PERMISSION -> "Permissions insuffisantes"
            ErrorType.NOT_FOUND -> "Ressource non trouvée"
            ErrorType.VALIDATION -> "Données invalides"
            ErrorType.UNKNOWN -> "Une erreur inattendue s'est produite"
        }
    }

    /**
     * Obtient l'historique des erreurs
     */
    fun getErrorHistory(): List<ErrorInfo> {
        return errorHistory.toList()
    }

    /**
     * Réinitialise l'historique
     */
    fun clearHistory() {
        errorHistory.clear()
    }

    companion object {
        private const val TAG = "ErrorHandler"
    }
}

/**
 * Type d'erreur
 */
enum class ErrorType {
    NETWORK,
    TIMEOUT,
    IO,
    AUTHENTICATION,
    PERMISSION,
    NOT_FOUND,
    VALIDATION,
    UNKNOWN
}

/**
 * Informations sur une erreur
 */
data class ErrorInfo(
    val type: ErrorType,
    val message: String,
    val stackTrace: String,
    val context: ErrorContext,
    val timestamp: Long
)

/**
 * Contexte d'une erreur
 */
data class ErrorContext(
    val screen: String? = null,
    val action: String? = null,
    val userId: String? = null,
    val additionalData: Map<String, Any> = emptyMap()
)

/**
 * Composable pour obtenir un ErrorHandler
 */
@Composable
fun rememberErrorHandler(
    scope: CoroutineScope = rememberCoroutineScope(),
    snackbarHostState: SnackbarHostState? = null
): ErrorHandler {
    return remember {
        ErrorHandler(scope, snackbarHostState)
    }
}


